### DISCUSSION TITLE
How do I render a file but not include it as a chapter in a book project?

### DISCUSSION THREAD

**Discussion by HarunCelikOtto** [quarto-dev/quarto-cli#10678] [Q&A]
### Description

Hello, 

I'm wondering if it is possible to have quarto render a `qmd` file in a book project without it actually being a part of the chapters of the book.

The idea is that as I am developing the book, I want to repurpose some of my writing into another file using the same `bib` file and `images\` directory.  I basically want to take advantage of Quarto assigning the project as root directory for files which are part of the project but not going to be rendered with the full book.

Assuming that my project directory and my `_quarto.yml` file look like the following.
```bash
Render-Testing
├── _quarto.yml
├── book-chapters
│   ├── chapter-one
│       └── ch1-sect-1.qmd
├── images
│   ├── chapter-one-imgs
│       └── methods.png
├── index.qmd
├── intro.qmd
├── references.bib
└── references.qmd
```

```yaml
project:
  type: book
  render:
    - "/book-chapters/chapter-one/ch1-sect-1.qmd"

book:
  title: "Render-Testing"
  author: "Norah Jones"
  date: "8/19/2024"
  chapters:
    - index.qmd
    - intro.qmd
    - references.qmd

bibliography: references.bib

format:
  html:
    theme: cosmo
  pdf:
    documentclass: scrreprt
```

I would like to be able to render the contents of `book-chapters/chapter-one/ch1-sect-1.qmd`, by running the `quarto render` command. This should include the correct paths to the `bib` file as well as to the `images/`. Running `quarto render` will not render the document, only those indicated in the chapters will render.

Contents of `ch1-sect-1.qmd`.
````qmd
---
title: Just Some Testing
---

Testing citations with [@knuth84]

Also some images.

![Test Image](/images/chapter-one-imgs/methods.png)
````

Output where the qmd file is specified in `quarto render`.
```
 quarto render  book-chapters/chapter-one/ch1-sect-1.qmd 
 quarto preview book-chapters/chapter-one/ch1-sect-1.qmd
```

![Screenshot from 2024-09-01 20-41-17](https://github.com/user-attachments/assets/f0837272-4099-4095-a8ae-3057bbc7bf37)

**Comment by mcanouil** [comment]
As we cannot use your code, it's hard to reply.

1. Create a book

```sh
quarto create project book
```

2. Remove `- summary.qmd` from `_quarto.yml`

3. Render the book `quarto render`

4. Render the document not referenced `quarto render summary.qmd`

Everything works fine.

If you expect `quarto render` to render everything that is not how it works currently.

About `project.render` (https://quarto.org/docs/websites/#render-targets), I don't know if it is supported at all in books.
CC @cscheid

**Comment by jarad** [comment]
@HarunCelikOtto I feel like I just ran into this issue as well. [In this commit](https://github.com/jarad/SportsAnalytics/commit/8cce3d29383e8a91b2dc02ab5452740ff50c232f), I needed the macros.qmd file to be rendered, but it never render when using `quarto render`. If this file was added as a chapter, then the file was rendered. But, like you, I didn't want it included as a chapter. 

@mcanouil I was expecting quarto to render every qmd file in the project directory because [it says so here](https://quarto.org/docs/websites/#render-targets). I am realizing now that this is what quarto does for websites and not necessarily for books that both @HarunCelikOtto and I are using. The [documentation for books](https://quarto.org/docs/books/) does not include the same section on what we should expect to be rendered. And apparently books do not render all qmd files in the project directory. 

[This post](https://stackoverflow.com/questions/74237333/quarto-render-single-html-file-in-a-project-without-rendering-the-whole-project) seems to have workaround. But it would be nice to have something builtin to the structure of quarto books that would allow you to render (in a particular order) files that are not chapters similar to [what is available for websites](https://quarto.org/docs/websites/#render-targets).

**Reply by HarunCelikOtto** [reply]
I never found a good workaround for this and in fact switched over to a website format just so I could use render targets properly. The only way this really worked is if I had the `.qmd` file in the root directory and I called the `quarto render` command on that file individually. This is quite sub-optimal since you should only really need to call the render once.

When I moved the file into any other sub-directory, and called `quarto render sub-directory/ *.qmd` the root paths to images and extensions would not resolve correctly.

**Reply by mcanouil** [reply]
It's unclear what your issues are here.
A small reproducible example shared as a Git repository would help.

In Quarto, paths are resolved relatively to the document calling the paths. A leading `/` means relative to the root of the project where `_quarto.yml` is.
This should work and as far as I know, it does.

**Reply by mcanouil** [reply]
To be clear, stating "I have a problem" or "it does not work" do not make it easy for others to help you (this is not specific to Quarto or software).
We need your help in order to provide any assistance.

**Reply by HarunCelikOtto** [reply]
> It's unclear what your issues are here. A small reproducible example shared as a Git repository would help.
> 
> In Quarto, paths are resolved relatively to the document calling the paths. A leading `/` means relative to the root of the project where `_quarto.yml` is. This should work and as far as I know, it does.

I thought my reply to this thread stating that the root path issue was clear, I apologize. 

What I tried to illustrate with my first example is that in a book project, a leading `/` does not refer to the root of the project if the file is in a sub directory of the root.

As a reproducible example. I have a new quarto book project and the only things I am doing are adding a testing image called `test.png`, a new directory called `newdir` and a file inside called `new.qmd` so that the directory tree looks like the following.

test
├── _quarto.yml
├── cover.png
├── imgs
  │   └── test.png
├── index.qmd
├── intro.qmd
├── newdir
  │   └── new.qmd
├── references.bib
├── references.qmd
└── summary.qmd

The `new.qmd` file is not a chapter in the `_quarto.yml` file and contains the following text inside.

```qmd
Here is an image ![Some Image](/imgs/test.png)
``` 

My expectation is that when I run `quarto render newdir/new.qmd`, the file should render with the referenced image. Instead, the image doesn't render at all.

**Reply by mcanouil** [reply]
Sorry, but showing a tree is not a reproducible example.
A reproducible example is something that readers can use.
You can't expect/ask users to build up the directory tree, figure out the content of the files, the command to run and then expect them to see the same thing as you especially without knowing the version of Quarto CLI being used.

**Reply by HarunCelikOtto** [reply]
Here is a repository of the directory I was talking about.
https://github.com/HarunCelikOtto/Quarto_Book_Test

I am using Quarto CLI version 1.6.40

**Reply by mcanouil** [reply]
I'm not sure what you expect from `quarto render newdir/new.qmd` in a book project.
It's like you are asking to compile a single file as a book.

If you want to have independent single files, you should use profiles: https://quarto.org/docs/projects/profiles.html

For example: `_quarto-single.yml`

```yml
project:
  type: default

format:
  html:
    theme: quartz
```

Then render via:

```sh
quarto render newdir/new.qmd --profile single
```

**Reply by jgostick** [reply]
Hi @mcanouil, I think I have the same need as the other commenters on this thread, so let me try to ask the question in a different way.  

I have a bunch of qmd (and ipynb) files which contain code to generate figures, BUT I do not use the `include` or `embed` short-codes since I find this makes it difficult to control the figure sizes in the output (a book fwiw).  Instead, the code in the qmd/ipymb *saves* the figures as png's, which I then insert into the book using `![](media/file.png)`.  So I think what I need is something like a short-code which runs qmd/ipynb files, but does not embed or include the results.  Like, is there a way to mimick a short-code like `{{< run file.qmd>}}`?

**Reply by mcanouil** [reply]
> So I think what I need is something like a short-code which runs qmd/ipynb files, but does not embed or include the results. Like, is there a way to mimick a short-code like {{< run file.qmd>}}?

This is the embed shortcode.
As the size is computed using the parameters from the notebook, the issue with size will be exactly the same. (Stretching a figure does not change the real size.)
The only way for you to compute the proper resolution and physical size for your "main" document is to put the evaluation itself inside it and to set the parameters as you see fit.
Any other method will require of you to set those parameters independently of the main document (either different values or to duplicate the values and keep them in "sync").

If you don't care about stretching/shrinking the image resulting in various type of quality losses, then your approach and setting the output width/height is the way to go.
